### PR TITLE
Plumb Peer DNS-SD advertisement for TCP support into SessionSetup context

### LINKED ISSUE #29539: Plumb Peer DNS-SD advertisement for TCP support into SessionSetup context
### Feature description

Currently, the TCP support advertisement from the peer is not correctly consumed and stored. This needs to be parsed properly during operational session setup and stored in the session context(given that the TCP support advertisement is a bit field in the 1.3 spec). This would be used to set the PeerAddress type to TCP or otherwise, when attempting to establish CASE with the peer.

### Platform

core

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by hnnajh**
Currently, the TCP support advertisement from the peer is not correctly consumed and stored. This needs to be parsed properly during operational session setup and stored in the session context(given that the TCP support advertisement is a bit field in the 1.3 spec). This would be used to set the PeerAddress type to TCP or otherwise, when attempting to establish CASE with the peer.

Fixes https://github.com/project-chip/connectedhomeip/issues/29539

### CODE REVIEW

**Review by pidarped — CHANGES_REQUESTED** [review]

File: src/app/OperationalSessionSetup.cpp [hunk 1]
```diff
@@ -178,7 +178,7 @@ void OperationalSessionSetup::Connect(Callback::Callback<OnDeviceConnected> * on
     }
 }
 
-void OperationalSessionSetup::UpdateDeviceData(const Transport::PeerAddress & addr, const ReliableMessageProtocolConfig & config)
+void OperationalSessionSetup::UpdateDeviceData(const Transport::PeerAddress & addr, const ReliableMessageProtocolConfig & config, bool supportsTcp)
```
[the remark this answers is not in the document]
Comment (hnnajh) [reply src/app/OperationalSessionSetup.cpp]: I agree, i moved things a little bit here 
I am going to take care of that division between TCPClient and TCPServer as part of https://github.com/project-chip/connectedhomeip/issues/30294

File: src/lib/dnssd/TxtFields.cpp [hunk 2]
```diff
@@ -268,9 +263,12 @@ void FillNodeDataFromTxt(const ByteSpan & key, const ByteSpan & value, CommonRes
     case TxtFieldKey::kSessionActiveThreshold:
         nodeData.mrpRetryActiveThreshold = Internal::GetRetryActiveThreshold(value);
         break;
-    case TxtFieldKey::kTcpSupported:
-        nodeData.supportsTcp = Internal::MakeBoolFromAsciiDecimal(value);
+    case TxtFieldKey::kTcpSupported: {
+        uint8_t support = Internal::MakeU8FromAsciiDecimal(value);
+        nodeData.supportsTcpClient = (support >> 1 & 0x01) == 0x01;
+        nodeData.supportsTcpServer = (support >> 2 & 0x01) == 0x01;
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/TxtFields.cpp:269]: Please don't write code that makes the reader figure out the order of operations.  This should be doing things more like `(support & 0x01) != 0` or something.

File: examples/chip-tool/commands/discover/Commands.h [hunk 3]
```diff
@@ -53,7 +53,8 @@ class Resolve : public DiscoverCommand, public chip::AddressResolve::NodeListene
                         result.mrpRemoteConfig.mIdleRetransTimeout.count());
         ChipLogProgress(chipTool, "   MRP retry interval (active): %" PRIu32 "ms",
                         result.mrpRemoteConfig.mActiveRetransTimeout.count());
-        ChipLogProgress(chipTool, "   Supports TCP: %s", result.supportsTcp ? "yes" : "no");
+        ChipLogProgress(chipTool, "   Supports Client TCP: %s", result.supportsTcpClient ? "yes" : "no");
+        ChipLogProgress(chipTool, "   Supports Server TCP: %s", result.supportsTcpServer ? "yes" : "no");
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/discover/Commands.h:57]:
```suggestion
        ChipLogProgress(chipTool, "   Supports TCP Client: %s", result.supportsTcpClient ? "yes" : "no");
        ChipLogProgress(chipTool, "   Supports TCP Server: %s", result.supportsTcpServer ? "yes" : "no");
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/dnssd/TxtFields.cpp (hunk 2 above)
Comment (pidarped) [reply src/lib/dnssd/TxtFields.cpp:269]: +1. Recommend defining an enum with TCP_CLIENT = 1, TCP_SERVER = 2. Then check `(support & (1 << TCP_CLIENT) != 0` for client support and `(support & (1 << TCP_SERVER) != 0` for server support.

**Review by pidarped — APPROVED** [review]

**Review by pidarped — CHANGES_REQUESTED** [review]
I don't see where the TCP Client and Server support is easily incorporated/accessed in OperationalSessionSetup and EstablishSession stage. Just like the remote and local MRP parameters are available, the TCP support capabilities need to be made available therein. These coupled with the Large Payload flag would dictate the choice of TCP as the transport.

File: src/lib/dnssd/TxtFields.cpp [hunk 4]
```diff
@@ -268,9 +263,12 @@ void FillNodeDataFromTxt(const ByteSpan & key, const ByteSpan & value, CommonRes
     case TxtFieldKey::kSessionActiveThreshold:
         nodeData.mrpRetryActiveThreshold = Internal::GetRetryActiveThreshold(value);
         break;
-    case TxtFieldKey::kTcpSupported:
-        nodeData.supportsTcp = Internal::MakeBoolFromAsciiDecimal(value);
+    case TxtFieldKey::kTcpSupported: {
+        uint8_t support = Internal::MakeU8FromAsciiDecimal(value);
```
[the remark this answers is not in the document]
Comment (hnnajh) [reply src/lib/dnssd/TxtFields.cpp]: done

**Review by pidarped — APPROVED** [review]
Approving based on offline discussion of bringing in the changes of populating the OperationalSessionSetup with the TCP support flags in a subsequent PR.

**Comment by pidarped** [comment]
Closing stale PR which has been resolved by PR https://github.com/project-chip/connectedhomeip/pull/33481.

**Pull request closed without merging**